### PR TITLE
Check if column exists before executing upgrade/downgrade operations in select revisions

### LINKED ISSUE #13617: The context manager that is supposed to silence a migration error doesn't work for postgresql.
The context manager that is supposed to silence a migration error (#13595) doesn't work for postgresql. As per Mike Bayer (author of Alembic and SQLAlchemy), "when an integrity error is issued by Postgresql, such as attempting to insert a primary key that already exists, the ongoing transaction is then marked as invalid. If you try to do anything else in the transaction, you get that error." ([ref](https://groups.google.com/g/sqlalchemy/c/r43b26pjzpY/m/g0aoIiVBWmYJ); same in [psycopg docs](https://www.psycopg.org/docs/faq.html))

**To Reproduce**
1. Make sure galaxy is up-to-date (current dev branch: `source_metadata` column should exist in `workflow` table)
2. Delete rows from `alembic_version` table
3. Run `manage_db.sh upgrade`, observe error (this will only happen on postgresql)

```
(.venv) rivendell$ ./manage_db.sh upgrade
Activating virtualenv at .venv
Traceback (most recent call last):
  File "/home/sergey/0dev/galaxy/_galaxy/dev/.venv/lib64/python3.9/site-packages/sqlalchemy/engine/base.py", line 1808, in _execute_context
    self.dialect.do_execute(
  File "/home/sergey/0dev/galaxy/_galaxy/dev/.venv/lib64/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block


The above exception was the direct cause of the following exception:

Traceback (most recent call last):

...

  File "/home/sergey/0dev/galaxy/_galaxy/dev/.venv/lib64/python3.9/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.InternalError: (psycopg2.errors.InFailedSqlTransaction) current transaction is aborted, commands ignored until end of transaction block

[SQL: UPDATE alembic_version SET version_num='b182f655505f' WHERE alembic_version.version_num = 'e7b6dcb09efd']
(Background on this error at: https://sqlalche.me/e/14/2j85)
```

**Expected behavior**
The error should be ignored.

**Solution Thoughts**
Ideally, we'd like to rollback the transaction - but if we do that, Alembic will not proceed to update the version, defeating the purpose of silencing the error.

### PR DESCRIPTION
**Pull request by jdavcs**
Fixes #13617 and replaces #13595 (because we don't need to silence those errors anymore).

I intentionally did not move this check into the utils module where it could be applied to all such schema modifications by default: I think that while it is justified for edge cases like `source_metadata` (and any schema edits prior to 22.05), it would be safer to fail fast in all other cases.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
1. Make sure galaxy is up-to-date (current dev branch: source_metadata column should exist in workflow table)
2. Delete rows from alembic_version table
3. Run manage_db.sh upgrade, observe no error happening. If you run these steps on the current dev branch, #13617 will happen.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py [hunk 1]
```diff
@@ -20,12 +20,16 @@
...
-    table, column = "workflow", "source_metadata"
-    with ignore_add_column_error(table, column):
-        op.add_column(table, Column(column, JSONType))
+    if not column_exists(table_name, column_name):
+        op.add_column(table_name, Column(column_name, JSONType))
 
 
 def downgrade():
-    drop_column("workflow", "source_metadata")
+    if column_exists(table_name, column_name):
```
Comment (mvdbeek) [inline lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py]: Do we need the check here? If we wanted to the migration to be "symmetrical" I suppose it should be a no-op. But given that the sqlalchemy-migrate migrations don't really fail I think we can drop unconditionally.

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py (hunk 1 above)
Comment (jdavcs) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py]: I don't think we need it here, but maybe I'm wrong. Here's my reasoning: this would be executed only by alembic, and that means alembic *has been used to upgrade* to the version containing that column - so the column *must* exist. The revision script will fail if the column does *not* exists - which indicates that the database is in an inconsistent state, which is when, I think, we want it to fail loud and fast. So, it's non-symmetrical by design, at least following this logic.

Comment (mvdbeek) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py]: > The revision script will fail if the column does _not_ exists

I agree with what you said before this, but I don't see how the script will fail. Don't you need to remove the if statement here ?

File: lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py [hunk 2]
```diff
@@ -20,12 +20,16 @@
...
+        drop_column(table_name, column_name)
```
Comment (mvdbeek) [inline lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py]:
```suggestion
    drop_column(table_name, column_name)
```
like this ?

File: lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py (hunk 1 above)
Comment (jdavcs) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b182f655505f_add_workflow_source_metadata_column.py]: I think I misread the quoted code - I thought you were suggesting to *add* the check! I mistook the red for the green; now my reply makes no sense whatsoever! :laughing: Thank you for catching this and sorry for the confusion!
